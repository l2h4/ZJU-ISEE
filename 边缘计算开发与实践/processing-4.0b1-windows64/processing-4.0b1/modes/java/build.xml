<?xml version="1.0"?>
<project name="Java Mode" default="build">

  <property name="generated"
	    value="${basedir}/generated/processing/mode/java/preproc" />
  <mkdir dir="${generated}" />

  <property name="grammars"
	    value="${basedir}/src/processing/mode/java/preproc" />

  <property name="antlr_jar"
	    value="${basedir}/mode/antlr-4.7.2-complete.jar" />

  <property name="mode_jar"
            value="${basedir}/mode/JavaMode.jar" />

  <classloader taskname="antlr">
    <classpath path="${antlr_jar}" />
  </classloader>

  <target name="clean" description="Clean the build directories">
    <delete dir="bin" />
    <delete dir="bin-test" />
    <delete file="${mode_jar}" />
    <delete>
      <fileset dir="${generated}">
	<include name="*.java" />
	<include name="*.tokens" />
	<include name="*.txt" />
	<include name="*.g" />
	<include name="*.smap" />
	<include name="*.properties" />
      </fileset>
    </delete>
  </target>

  <target name="preproc" description="Compile ANTLR 4 grammar">
    <java classname="org.antlr.v4.Tool"
        failonerror="true">
      <classpath path="${antlr_jar}" />
      <arg value="-o" />
      <arg value="${generated}" />
      <arg value="-package" />
      <arg value="processing.mode.java.preproc" />
      <arg value="${grammars}/Processing.g4" />
    </java>
  </target>

  <path id="classpath.base">
    <pathelement location="../core/library/core.jar" />
    <pathelement location="../app/pde.jar" />
    <pathelement location="../app/lib/ant.jar" />
    <pathelement location="../app/lib/ant-launcher.jar" />
    <pathelement location="../app/lib/apple.jar" />
    <pathelement location="../app/lib/jna.jar" />
    <pathelement location="../app/lib/jna-platform.jar" />
    <pathelement location="../app/lib/VAqua7.jar" />

    <pathelement location="mode/antlr-4.7.2-complete.jar" />
    <pathelement location="mode/classpath-explorer-1.0.jar" />
    <pathelement location="mode/jsoup-1.7.1.jar" />
    <pathelement location="mode/org.netbeans.swing.outline.jar" />
    <pathelement location="mode/com.ibm.icu.jar" />

    <pathelement location="mode/org.eclipse.core.contenttype.jar" />
    <pathelement location="mode/org.eclipse.core.jobs.jar" />
    <pathelement location="mode/org.eclipse.core.resources.jar" />
    <pathelement location="mode/org.eclipse.core.runtime.jar" />
    <pathelement location="mode/org.eclipse.equinox.common.jar" />
    <pathelement location="mode/org.eclipse.equinox.preferences.jar" />
    <pathelement location="mode/org.eclipse.jdt.compiler.apt.jar" />
    <pathelement location="mode/org.eclipse.jdt.core.jar" />
    <pathelement location="mode/org.eclipse.osgi.jar" />
    <pathelement location="mode/org.eclipse.text.jar" />
  </path>

  <path id="classpath.test">
    <pathelement location="../core/library-test/junit-4.8.1.jar" />
    <pathelement location="../core/library-test/mockito-all-1.10.19.jar" />
    <pathelement location = "bin-test" />
    <path refid="classpath.base" />
  </path>

  <macrodef name="compilecommon">
    <attribute name="destdir"/>
    <attribute name="srcdir"/>
    <attribute name="classpath"/>
    <sequential>
      <condition property="core-built">
        <available file="../core/library/core.jar" />
      </condition>
      <fail unless="core-built" message="Please first build the core library and make sure it is located at ../core/library/core.jar" />

      <condition property="app-built">
        <available file="../app/pde.jar" />
      </condition>
      <fail unless="app-built" message="Please build app first and make sure it is located at ../app/pde.jar" />

      <mkdir dir="@{destdir}" />

      <!-- in some cases, pde.jar was not getting built
  	 https://github.com/processing/processing/issues/1792 -->
      <delete file="${mode_jar}" />

      <!-- env used to set classpath below -->
      <property environment="env" />

      <javac source="11"
  	   target="11"
  	   destdir="@{destdir}"
  	   excludes="**/tools/format/**"
  	   encoding="UTF-8"
  	   includeAntRuntime="false"
  	   debug="on"
  	   nowarn="true"
  	   compiler="modern">
        <src path="@{srcdir}" />
        <src path="generated" />
        <classpath refid="@{classpath}"/>
      </javac>
    </sequential>
  </macrodef>

  <target name="test-compile" depends="preproc">
    <compilecommon srcdir="src; test/processing" destdir="bin-test" classpath="classpath.test" />
  </target>

  <target name="test" depends="build, test-compile">
    <junit haltonfailure="true">
      <classpath refid="classpath.test" />
      <formatter type="brief" usefile="false" />
      <batchtest>
        <fileset dir="test">
          <include name="**/*Test.java" />
          <include name="**/*Tests.java" />
        </fileset>
      </batchtest>
    </junit>
  </target>

  <target name="compile" description="Compile sources" depends="preproc">
    <compilecommon srcdir="src" destdir="bin" classpath="classpath.base" />
  </target>

  <target name="build" depends="compile" description="Build PDE">
    <jar basedir="bin" destfile="${mode_jar}" />
  </target>

</project>
