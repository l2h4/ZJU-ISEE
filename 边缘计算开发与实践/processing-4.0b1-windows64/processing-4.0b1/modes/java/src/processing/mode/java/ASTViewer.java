package processing.mode.java;

import java.awt.EventQueue;
import java.awt.Rectangle;
import java.awt.event.ComponentAdapter;
import java.awt.event.ComponentEvent;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.function.Consumer;

import javax.swing.JDialog;
import javax.swing.JScrollPane;
import javax.swing.JTree;
import javax.swing.WindowConstants;
import javax.swing.tree.DefaultMutableTreeNode;
import javax.swing.tree.DefaultTreeModel;

import org.eclipse.jdt.core.dom.ASTNode;
import org.eclipse.jdt.core.dom.ASTVisitor;
import org.eclipse.jdt.core.dom.CompilationUnit;

import processing.app.Messages;
import processing.app.ui.ZoomTreeCellRenderer;


class ASTViewer {
  final JDialog window;
  final JTree tree;
  final Consumer<PreprocSketch> updateListener;


  ASTViewer(JavaEditor editor, PreprocService pps) {
    updateListener = this::buildAndUpdateTree;

    window = new JDialog(editor);

    tree = new JTree() {
      @Override
      public String convertValueToText(Object value, boolean selected,
                                       boolean expanded, boolean leaf,
                                       int row, boolean hasFocus) {
        if (value instanceof DefaultMutableTreeNode) {
          DefaultMutableTreeNode treeNode = (DefaultMutableTreeNode) value;
          Object o = treeNode.getUserObject();
          if (o instanceof ASTNode) {
            ASTNode node = (ASTNode) o;
            return CompletionGenerator.getNodeAsString(node);
          }
        }
        return super.convertValueToText(value, selected, expanded, leaf, row, hasFocus);
      }
    };
    tree.setCellRenderer(new ZoomTreeCellRenderer());
    window.addComponentListener(new ComponentAdapter() {
      @Override
      public void componentHidden(ComponentEvent e) {
        pps.unregisterListener(updateListener);
        tree.setModel(null);
      }
    });
    window.setDefaultCloseOperation(WindowConstants.HIDE_ON_CLOSE);
    window.setBounds(new Rectangle(680, 100, 460, 620));
    window.setTitle("AST View - " + editor.getSketch().getName());
    JScrollPane sp = new JScrollPane();
    sp.setViewportView(tree);
    window.add(sp);
    pps.whenDone(updateListener);
    pps.registerListener(updateListener);

    tree.addTreeSelectionListener(e -> {
      if (tree.getLastSelectedPathComponent() != null) {
        DefaultMutableTreeNode treeNode =
          (DefaultMutableTreeNode) tree.getLastSelectedPathComponent();
        if (treeNode.getUserObject() instanceof ASTNode) {
          ASTNode node = (ASTNode) treeNode.getUserObject();
          pps.whenDone(ps -> {
            SketchInterval si = ps.mapJavaToSketch(node);
            if (!ps.inRange(si)) return;
            EventQueue.invokeLater(() -> editor.highlight(si.tabIndex, si.startTabOffset, si.stopTabOffset));
          });
        }
      }
    });
  }


  void dispose() {
    if (window != null) {
      window.dispose();
    }
  }


  // Thread: worker
  void buildAndUpdateTree(PreprocSketch ps) {
    CompilationUnit cu = ps.compilationUnit;
    if (cu.types().isEmpty()){
      Messages.err("No Type found in CU");
      return;
    }

    Deque<DefaultMutableTreeNode> treeNodeStack = new ArrayDeque<>();

    ASTNode type0 = (ASTNode) cu.types().get(0);
    type0.accept(new ASTVisitor() {
      @Override
      public boolean preVisit2(ASTNode node) {
        treeNodeStack.push(new DefaultMutableTreeNode(node));
        return super.preVisit2(node);
      }

      @Override
      public void postVisit(ASTNode node) {
        if (treeNodeStack.size() > 1) {
          DefaultMutableTreeNode treeNode = treeNodeStack.pop();
          treeNodeStack.peek().add(treeNode);
        }
      }
    });

    DefaultMutableTreeNode codeTree = treeNodeStack.pop();

    EventQueue.invokeLater(() -> {
      if (tree.hasFocus() || window.hasFocus()) {
        return;
      }
      tree.setModel(new DefaultTreeModel(codeTree));
      ((DefaultTreeModel) tree.getModel()).reload();
      tree.validate();
      if (!window.isVisible()) {
        window.setVisible(true);
      }
    });
  }
}