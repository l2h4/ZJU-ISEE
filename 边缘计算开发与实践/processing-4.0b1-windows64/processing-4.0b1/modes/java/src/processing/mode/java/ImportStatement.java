/* -*- mode: java; c-basic-offset: 2; indent-tabs-mode: nil -*- */

/*
Part of the Processing project - http://processing.org
Copyright (c) 2012-19 The Processing Foundation

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License version 2
as published by the Free Software Foundation.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software Foundation, Inc.
51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
*/

package processing.mode.java;

import java.util.regex.MatchResult;
import java.util.regex.Matcher;

/**
 * Wrapper for import statements
 */
public class ImportStatement {

  private static final String importKw = "import";
  private static final String staticKw = "static";

//  private boolean isClass;
  private boolean isStarred;
  private boolean isStatic;

  /**
   * Full class name of the import with all packages
   * Ends with star for starred imports
   */
  private String memberName;

  /**
   * Name of the package e.g. everything before last dot
   */
  private String packageName;

  private ImportStatement() { }

  /**
   * Create an import statement for a full package.
   *
   * @param cls The fully qualified name of the package.
   * @return ImportStatement which imports all package members in a non-static context using a wildcard.
   */
  public static ImportStatement wholePackage(String pckg) {
    ImportStatement is = new ImportStatement();
    is.packageName = pckg;
    is.memberName = "*";
    is.isStarred = true;
    return is;
  }

  /**
   * Create an import statement for a single class.
   *
   * @param cls The fully qualified name of the class.
   * @return ImportStatement which imports the class in a non-static context.
   */
  public static ImportStatement singleClass(String cls) {
    ImportStatement is = new ImportStatement();
    int lastDot = cls.lastIndexOf('.');
    is.memberName = lastDot >= 0 ? cls.substring(lastDot+1) : cls;
    is.packageName = lastDot >= 0 ? cls.substring(0, lastDot) : "";
//    is.isClass = true;
    return is;
  }

  /**
   * Prase an import statement from a fully qualified name.
   *
   * @param importString The fully qualified name from which an import statement should be built. This supports static
   *    prepended so both "java.util.List" and "static org.processing.package.Factory.build" are supported but
   *    "import java.util.List;" is not. Note that the static prepending is required if the import is static.
   * @return Newly parsed import statement information.
   */
  public static ImportStatement parse(String importString) {
    Matcher matcher = SourceUtil.IMPORT_REGEX_NO_KEYWORD.matcher(importString);
    if (!matcher.find()) return null;

    return parse(matcher.toMatchResult());
  }

  /**
   * Parse an import statement from a regex match found via SourceUtils.*REGEX*.
   *
   * @param match The regex match from which an import statement should be built. Can be from IMPORT_REGEX_NO_KEYWORD or
   *    IMPORT_REGEX or equivalent.
   * @return Newly parsed import statement information.
   */
  public static ImportStatement parse(MatchResult match) {
    ImportStatement is = new ImportStatement();

    is.isStatic = match.group(2) != null;
    String pckg = match.group(3);
    pckg = (pckg == null) ? "" : pckg.replaceAll("\\s","");

    String memberName = match.group(4);
    is.isStarred = memberName.equals("*");

    // Deal with static member imports whose "className" is actually a "member name".
    boolean endsWithPeriod = pckg.endsWith(".");
    if (is.isStatic) {
      String withContainingTypeNameAtEnd = endsWithPeriod ? pckg.substring(0, pckg.length() - 1) : pckg;
      int periodOfContainingTypeName = withContainingTypeNameAtEnd.lastIndexOf(".");
      String containingTypeName = withContainingTypeNameAtEnd.substring(periodOfContainingTypeName + 1);
      is.packageName = withContainingTypeNameAtEnd.substring(0, periodOfContainingTypeName);
      is.memberName = containingTypeName + "." + memberName;
    } else {
      is.packageName = endsWithPeriod ? pckg.substring(0, pckg.length() - 1) : pckg;
      is.memberName = memberName;
    }

    return is;
  }

  /**
   * Get the source line needed to execute this import.
   *
   * @return The java code required for executing this import.
   */
  public String getFullSourceLine() {
    return importKw + " " + (isStatic ? (staticKw + " ") : "") + packageName + "." + memberName + ";";
  }

  /**
   * Get the fully qualified member name which includes the package path.
   *
   * @return The fully qualified member name including the parent class. This is "java.util.List" in the case of
   *    "import java.util.List". Note that, in the case of static imports, it will include the member imported so
   *    "org.processing.package.Factory.build" would be returned for
   *    "import static org.processing.package.Factory.build".
   */
  public String getFullMemberName(){
    return packageName + "." + memberName;
  }

  /**
   * Get the end of the import statement with the type to be imported.
   *
   * @return This is the class name (List) in the case of "import java.util.List" or "*" in the case of a wildcard. For
   *    static imports this will be the member within the containing class (Factory.build in the case of
   *    "import static org.processing.package.Factory.build").
   */
  public String getMemberName(){
    return memberName;
  }

  /**
   * Get the package from which the import is being made.
   *
   * @return The package "java.util" from "import java.util.List". Note that, in the case of wildcards, the wildcard
   *    will be in the member name and not the class name.
   */
  public String getPackageName(){
    return packageName;
  }

  /**
   * Determine if this import is a wildcard import.
   *
   * @return True if the FQN (fully qualified name) ends in a wildcard and false otherwise.
   */
  public boolean isStarredImport() {
    return isStarred;
  }

  /**
   * Determine if this import statement is a static import.
   *
   * @return True if of the form "import static {FQN}" where FQN refers to the fully qualified name and false otherwise.
   */
  public boolean isStaticImport() {
    return isStatic;
  }

  /**
   * Check if the import statements refer to the same import target.
   *
   * @param is The other import statement.
   * @return True of the two ImportStatements refer to the same import target and false otherwise.
   */
  public boolean isSameAs(ImportStatement is) {
    return packageName.equals(is.packageName) &&
        memberName.equals(is.memberName) &&
        isStatic == is.isStatic;
  }
}