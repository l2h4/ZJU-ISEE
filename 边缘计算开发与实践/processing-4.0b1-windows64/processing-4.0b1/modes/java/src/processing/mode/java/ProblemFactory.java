package processing.mode.java;

import java.util.List;
import java.util.OptionalInt;
import java.util.stream.IntStream;

import processing.app.Problem;
import processing.app.ui.Editor;
import processing.mode.java.preproc.PdePreprocessIssue;


/**
 * Factory which helps create {Problem}s during preprocessing.
 */
public class ProblemFactory {

  /**
   * Create a new {Problem}.
   *
   * @param pdePreprocessIssue The preprocess issue found.
   * @param tabStarts The list of line numbers on which each tab starts.
   * @param editor The editor in which errors will appear.
   * @return Newly created problem.
   */
  public static Problem build(PdePreprocessIssue pdePreprocessIssue, List<Integer> tabStarts,
      int numLines, Editor editor) {

    int line = pdePreprocessIssue.getLine();

    // Sometimes errors are reported one line past end of sketch. Fix that.
    if (line >= numLines) {
      line = numLines - 1;
    }

    // Get local area
    TabLine tabLine = getTab(tabStarts, line);

    int tab = tabLine.getTab();
    int localLine = tabLine.getLineInTab(); // Problems emitted in 0 index

    // Generate syntax problem
    String message = pdePreprocessIssue.getMsg();

    int lineStart = editor.getLineStartOffset(localLine);
    int lineStop = editor.getLineStopOffset(localLine) - 1;

    if (lineStart == lineStop) {
      lineStop++;
    }

    return new SyntaxProblem(
        tab,
        localLine,
        message,
        lineStart,
        lineStop
    );
  }

  /**
   * Create a new {Problem}.
   *
   * @param pdePreprocessIssue The preprocess issue found.
   * @param tabStarts The list of line numbers on which each tab starts.
   * @return Newly created problem.
   */
  public static Problem build(PdePreprocessIssue pdePreprocessIssue, List<Integer> tabStarts) {
    int line = pdePreprocessIssue.getLine();

    TabLine tabLine = getTab(tabStarts, line);

    int tab = tabLine.getTab();
    int localLine = tabLine.getLineInTab();
    int col = pdePreprocessIssue.getCharPositionInLine();

    String message = pdePreprocessIssue.getMsg();

    if (col == 0) {
      col = 1;
    }

    return new SyntaxProblem(
        tab,
        localLine,
        message,
        localLine,
        localLine + col
    );
  }

  /**
   * Get the local tab and line number for a global line.
   *
   * @param tabStarts The lines on which each tab starts.
   * @param line The global line to locate as a local line.
   * @return The local tab number and local line number.
   */
  protected static TabLine getTab(List<Integer> tabStarts, int line) {
    OptionalInt tabMaybe = IntStream.range(0, tabStarts.size())
        .filter((index) -> line >= tabStarts.get(index))
        .max();

    int tab = tabMaybe.orElse(0);

    int localLine = line - tabStarts.get(tab);

    return new TabLine(tab, line, localLine);
  }

}
