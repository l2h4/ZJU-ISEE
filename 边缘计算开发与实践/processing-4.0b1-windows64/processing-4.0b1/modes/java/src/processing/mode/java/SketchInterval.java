package processing.mode.java;

public class SketchInterval {

  public static final SketchInterval BEFORE_START = new SketchInterval(-1, -1, -1, -1, -1);

  SketchInterval(int tabIndex,
                         int startTabOffset, int stopTabOffset,
                         int startPdeOffset, int stopPdeOffset) {
    this.tabIndex = tabIndex;
    this.startTabOffset = startTabOffset;
    this.stopTabOffset = stopTabOffset;
    this.startPdeOffset = startPdeOffset;
    this.stopPdeOffset = stopPdeOffset;
  }

  final int tabIndex;
  final int startTabOffset;
  final int stopTabOffset;

  final int startPdeOffset;
  final int stopPdeOffset;
}