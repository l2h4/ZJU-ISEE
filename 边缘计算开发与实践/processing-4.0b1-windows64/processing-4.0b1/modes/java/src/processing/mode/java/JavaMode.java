/* -*- mode: java; c-basic-offset: 2; indent-tabs-mode: nil -*- */

/*
  Part of the Processing project - http://processing.org

  Copyright (c) 2010-11 Ben Fry and Casey Reas
  Copyright (c) 2012-19 The Processing Foundation

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License version 2
  as published by the Free Software Foundation.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software Foundation,
  Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

package processing.mode.java;

import java.awt.*;
import java.io.File;
import java.io.IOException;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

import processing.app.*;
import processing.app.ui.Editor;
import processing.app.ui.EditorException;
import processing.app.ui.EditorState;

import processing.mode.java.runner.Runner;
import processing.mode.java.tweak.SketchParser;


public class JavaMode extends Mode {

  public Editor createEditor(Base base, String path,
                             EditorState state) throws EditorException {
    return new JavaEditor(base, path, state, this);
  }


  public JavaMode(Base base, File folder) {
    super(base, folder);

    loadPreferences();
    loadSuggestionsMap();
  }


  public String getTitle() {
    return "Java";
  }


  // . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . .


  public File[] getExampleCategoryFolders() {
    return new File[] {
      new File(examplesFolder, "Basics"),
      new File(examplesFolder, "Topics"),
      new File(examplesFolder, "Demos"),
      new File(examplesFolder, "Books")
    };
  }


  public String getDefaultExtension() {
    return "pde";
  }


  public String[] getExtensions() {
    return new String[] { "pde", "java" };
  }


  public String[] getIgnorable() {
    return new String[] {
      "applet",
      "application.macosx",
      "application.windows",
      "application.linux"
    };
  }


  public Library getCoreLibrary() {
    if (coreLibrary == null) {
      File coreFolder = Platform.getContentFile("core");
      coreLibrary = new Library(coreFolder);
    }
    return coreLibrary;
  }


  // . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . .



  /** Handles the standard Java "Run" or "Present" */
  public Runner handleLaunch(Sketch sketch, RunnerListener listener,
                             final boolean present) throws SketchException {
    JavaBuild build = new JavaBuild(sketch);
//    String appletClassName = build.build(false);
    String appletClassName = build.build(true);
    if (appletClassName != null) {
      final Runner runtime = new Runner(build, listener);
      new Thread(() -> {
        // these block until finished
        if (present) {
          runtime.present(null);
        } else {
          runtime.launch(null);
        }
      }).start();
      return runtime;
    }
    return null;
  }


  /** Start a sketch in tweak mode */
  public Runner handleTweak(Sketch sketch,
                            RunnerListener listener,
                            JavaEditor editor) throws SketchException {
    // first try to build the unmodified code
    JavaBuild build = new JavaBuild(sketch);
//    String appletClassName = build.build(false);
    String appletClassName = build.build(true);
    if (appletClassName == null) {
      // unmodified build failed, so fail
      return null;
    }

    // if compilation passed, modify the code and build again
    // save the original sketch code of the user
    editor.initBaseCode();
    // check for "// tweak" comment in the sketch
    boolean requiresTweak = SketchParser.containsTweakComment(editor.baseCode);
    // parse the saved sketch to get all (or only with "//tweak" comment) numbers
    final SketchParser parser = new SketchParser(editor.baseCode, requiresTweak);

    // add our code to the sketch
    final boolean launchInteractive = editor.automateSketch(sketch, parser);

    build = new JavaBuild(sketch);
    appletClassName = build.build(false);

    if (appletClassName != null) {
      final Runner runtime = new Runner(build, listener);
      new Thread(() -> {
        runtime.launch(null);
        // next lines are executed when the sketch quits
        if (launchInteractive) {
          // fix swing deadlock issue: https://github.com/processing/processing/issues/3928
          EventQueue.invokeLater(() -> {
            editor.initEditorCode(parser.allHandles, false);
            editor.stopTweakMode(parser.allHandles);
          });
        }
      }).start();

      if (launchInteractive) {
        // fix swing deadlock issue: https://github.com/processing/processing/issues/3928
        EventQueue.invokeLater(() -> {
          // replace editor code with baseCode
          editor.initEditorCode(parser.allHandles, false);
          editor.updateInterface(parser.allHandles, parser.colorBoxes);
          editor.startTweakMode();
        });
      }
      return runtime;
    }
    return null;
  }


  public boolean handleExportApplication(Sketch sketch) throws SketchException, IOException {
    JavaBuild build = new JavaBuild(sketch);
    return build.exportApplication();
  }


  /**
   * Any modes that extend JavaMode can override this method to add additional
   * JARs to be included in the classpath for code completion and error checking
   * @return searchPath: file-paths separated by File.pathSeparatorChar
   */
  public String getSearchPath() {
    return getCoreLibrary().getJarPath();
  }


  // . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . .

  // Merged from ExperimentalMode


  static volatile boolean errorCheckEnabled = true;
  static volatile boolean warningsEnabled = true;
  static volatile boolean errorLogsEnabled = false;

  static volatile boolean autoSaveEnabled = true;
  static volatile boolean autoSavePromptEnabled = true;
  static volatile boolean defaultAutoSaveEnabled = true;

  static volatile boolean codeCompletionsEnabled = true;
  static volatile boolean ccTriggerEnabled = false;
  static volatile boolean importSuggestEnabled = true;
  static volatile boolean inspectModeHotkeyEnabled = true;

  static public int autoSaveInterval = 3; //in minutes

  static public final String prefErrorCheck = "pdex.errorCheckEnabled";
  static public final String prefWarnings = "pdex.warningsEnabled";
  static public final String prefErrorLogs = "pdex.writeErrorLogs";

  static public final String prefAutoSave = "pdex.autoSave.autoSaveEnabled";
  static public final String prefAutoSaveInterval = "pdex.autoSaveInterval";
  static public final String prefAutoSavePrompt = "pdex.autoSave.promptDisplay";
  static public final String prefDefaultAutoSave = "pdex.autoSave.autoSaveByDefault";

  static public final String COMPLETION_PREF = "pdex.completion";
  static public final String COMPLETION_TRIGGER_PREF = "pdex.completion.trigger";
  static public final String SUGGEST_IMPORTS_PREF = "pdex.suggest.imports";
  static public final String INSPECT_MODE_HOTKEY_PREF = "pdex.inspectMode.hotkey";

  /**
   * Stores the white list/black list of allowed/blacklisted imports.
   * These are defined in suggestions.txt in java mode folder.
   */
  private final Set<String> includedSuggestions = ConcurrentHashMap.newKeySet();
  private final Set<String> excludedSuggestions = ConcurrentHashMap.newKeySet();


  boolean includeSuggestion(String impName) {
    return includedSuggestions.contains(impName);
  }


  boolean excludeSuggestion(String impName) {
    return excludedSuggestions.contains(impName);
  }


  /*
  static boolean checkSuggestion(String mapName, String impName) {
    return suggestionsMap.containsKey(mapName) &&
      suggestionsMap.get(mapName).contains(impName);
  }
  */


  public void loadPreferences() {
    Messages.log("Load PDEX prefs");

    errorCheckEnabled = Preferences.getBoolean(prefErrorCheck);
    warningsEnabled = Preferences.getBoolean(prefWarnings);
    errorLogsEnabled = Preferences.getBoolean(prefErrorLogs);

    autoSaveEnabled = Preferences.getBoolean(prefAutoSave);
    autoSaveInterval = Preferences.getInteger(prefAutoSaveInterval);
    autoSavePromptEnabled = Preferences.getBoolean(prefAutoSavePrompt);
    defaultAutoSaveEnabled = Preferences.getBoolean(prefDefaultAutoSave);

    codeCompletionsEnabled = Preferences.getBoolean(COMPLETION_PREF);
    ccTriggerEnabled = Preferences.getBoolean(COMPLETION_TRIGGER_PREF);
    importSuggestEnabled = Preferences.getBoolean(SUGGEST_IMPORTS_PREF);
    inspectModeHotkeyEnabled = Preferences.getBoolean(INSPECT_MODE_HOTKEY_PREF);
  }


  public void savePreferences() {
    Messages.log("Saving PDEX prefs");

    Preferences.setBoolean(prefErrorCheck, errorCheckEnabled);
    Preferences.setBoolean(prefWarnings, warningsEnabled);
    Preferences.setBoolean(prefErrorLogs, errorLogsEnabled);

    Preferences.setBoolean(prefAutoSave, autoSaveEnabled);
    Preferences.setInteger(prefAutoSaveInterval, autoSaveInterval);
    Preferences.setBoolean(prefAutoSavePrompt, autoSavePromptEnabled);
    Preferences.setBoolean(prefDefaultAutoSave, defaultAutoSaveEnabled);

    Preferences.setBoolean(COMPLETION_PREF, codeCompletionsEnabled);
    Preferences.setBoolean(COMPLETION_TRIGGER_PREF, ccTriggerEnabled);
    Preferences.setBoolean(SUGGEST_IMPORTS_PREF, importSuggestEnabled);
    Preferences.setBoolean(INSPECT_MODE_HOTKEY_PREF, inspectModeHotkeyEnabled);
  }


  private void loadSuggestionsMap() {
    Collections.addAll(includedSuggestions, getSuggestionIncludeList());
    Collections.addAll(excludedSuggestions, getSuggestionExcludeList());
  }


  // broken out so that it can be overridden by Android, etc
  protected String[] getSuggestionIncludeList() {
    return new String[] {
      "processing.core.PApplet",
      "processing.core.PFont",
      "processing.core.PGraphics",
      "processing.core.PImage",
      "processing.core.PMatrix2D",
      "processing.core.PMatrix3D",
      "processing.core.PStyle",
      "processing.core.PVector",
      "processing.core.PShape",
      "processing.core.PGraphicsJava2D",
      "processing.core.PGraphics2D",
      "processing.core.PGraphics3D",
      "processing.data.FloatDict",
      "processing.data.FloatList",
      "processing.data.IntDict",
      "processing.data.IntList",
      "processing.data.JSONArray",
      "processing.data.JSONObject",
      "processing.data.StringDict",
      "processing.data.StringList",
      "processing.data.Table",
      "processing.data.XML",
      "processing.event.Event",
      "processing.event.KeyEvent",
      "processing.event.MouseEvent",
      "processing.event.TouchEvent",
      "processing.opengl.PShader",
      "processing.opengl.PGL",

      "java.util.ArrayList",
      "java.io.BufferedReader",
      "java.util.HashMap",
      "java.io.PrintWriter",
      "java.lang.String"
    };
  }


  // broken out so that it can be overridden by Android, etc
  protected String[] getSuggestionExcludeList() {
    return new String[] {
      "processing.core.PGraphicsRetina2D",
      "processing.core.PShapeOBJ",
      "processing.core.PShapeSVG",
      "processing.data.Sort",
      "processing.opengl.FrameBuffer",
      "processing.opengl.LinePath",
      "processing.opengl.LinePath.PathIterator",
      "processing.opengl.LineStroker",
      "processing.opengl.PGraphicsOpenGL"
    };
  }


  /*
  private void loadSuggestionsMap() {
    File suggestionsFile = new File(getFolder(), "suggestions.txt");
    if (suggestionsFile.exists()) {
      String[] lines = PApplet.loadStrings(suggestionsFile);
      if (lines != null) {
        for (String line : lines) {
          line = line.trim();
          if (line.length() > 0 && !line.startsWith("#")) {
            int equals = line.indexOf('=');
            if (equals != -1) {
              String key = line.substring(0, equals).trim();
              String value = line.substring(equals + 1).trim();

              if (key.equals("include")) {
                includedSuggestions.add(value);
              } else if (key.equals("exclude")) {
                excludedSuggestions.add(value);
              } else {
                Messages.loge("Should be include or exclude: " + key);
              }
            } else {
              Messages.loge("Bad line found in suggestions file: " + line);
            }
          }
        }
      }
    } else {
      Messages.loge("Suggestions file not found at " + suggestionsFile);
    }
  }
  */
}
