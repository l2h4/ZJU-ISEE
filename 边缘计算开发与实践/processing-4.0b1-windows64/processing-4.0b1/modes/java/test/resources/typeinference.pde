import java.util.*;
import java.util.function.*;

void setup() {
  List<String> list = new ArrayList<>();
  list.add("line1\nline2");
  list.add("line3");

  // Local variable type inference in loop
  for (var s : list) {
    println(s);
  }

  // Local variable type inference
  var testString = list.get(0);
  println(testString.lines().count()); // Java 11 API
}
