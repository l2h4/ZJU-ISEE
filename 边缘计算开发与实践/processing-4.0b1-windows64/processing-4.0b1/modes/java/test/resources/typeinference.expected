import processing.core.*;
import processing.data.*;
import processing.event.*;
import processing.opengl.*;

import java.util.*;
import java.util.function.*;

import java.util.HashMap;
import java.util.ArrayList;
import java.io.File;
import java.io.BufferedReader;
import java.io.PrintWriter;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.IOException;

public class typeinference extends PApplet {

  public void setup() {
    List<String> list = new ArrayList<>();
    list.add("line1\nline2");
    list.add("line3");

    // Local variable type inference in loop
    for (var s : list) {
      println(s);
    }

    // Local variable type inference
    var testString = list.get(0);
    println(testString.lines().count()); // Java 11 API
  }


  static public void main(String[] passedArgs) {
    String[] appletArgs = new String[] { "typeinference" };
    if (passedArgs != null) {
      PApplet.main(concat(appletArgs, passedArgs));
    } else {
      PApplet.main(appletArgs);
    }
  }
}
